genie-web/src/main/java/com/netflix/genie/web/agent/apis/rpc/v4/endpoints/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -142,333 +151,630 @@ public GRpcAgentFileStreamServiceImpl(
             }
+        }
 
-            if (!messageStreamId.equals(streamId.get())) {
-                log.warn(
-                    "Received chunk with id: {}, but this stream was previously used by stream: {}",
-                    messageStreamId,
-                    streamId.get()
-                );
-                return;
+        private synchronized FileTransfer findFileTransfer(final AgentFileChunkObserver agentFileChunkObserver) {
+            // Find a transfer by iterating over the active ones.
+            // Could keep a map indexed by observers, but this should be fine since it's only called when a stream
+            // terminates.
+            for (final FileTransfer fileTransfer : this.activeTransfers.values()) {
+                if (fileTransfer.getAgentFileChunkObserver() == agentFileChunkObserver) {
+                    return fileTransfer;
+                }
             }
+            return null;
+        }
 
-            // May block if the queue of chunks is full
-            this.gRpcAgentFileStreamService.handleFileTransferChunk(
-                this,
-                value.getStreamId(),
-                value.getData()
-            );
+        // N.B. this should not synchronized to avoid locking up the transfer manager
+        private void writeDataAndAck(final FileTransfer fileTransfer, final ByteString data) {
+            final String fileTransferId = fileTransfer.getTransferId();
+            try {
+                // Try to write. May fail if buffer consumer is slow and buffer is not drained yet.
+                if (fileTransfer.append(data)) {
+                    log.debug("Wrote chunk of transfer {} to buffer. Sending ack", fileTransferId);
+                    fileTransfer.sendAck();
+                } else {
+                    // Try again in a little bit
+                    this.taskScheduler.schedule(
+                        () -> this.writeDataAndAck(fileTransfer, data),
+                        Instant.now().plus(this.properties.getWriteRetryDelay())
+                    );
+                }
+            } catch (IllegalStateException e) {
+                // Eventually retries will stop because the transfer times out due to lack of progress
+                log.warn("Buffer of transfer {} is closed", fileTransferId);
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public synchronized void contribute(final Info.Builder builder) {
+            builder
+                .withDetail(
+                    "active_file_transfers",
+                    this.activeTransfers.values()
+                        .stream()
+                        .map(FileTransfer::toString)
+                        .collect(Collectors.toList())
+                )
+                .withDetail("unclaimed_transfer_streams", this.unclaimedTransferStreams.size());
+        }
+    }
+
+    private static final class FileTransfer {
```

Comment:
There's a bunch of private classes in this file. Might help readability to break them out into perhaps package protected independent classes?

Reply:
Yeah this file is long and this class is complex. It's juggling gRPC streams and reacting to different asynchronous events (stream errors, messages, download cancelled, timeouts, ...).